TMEDIA-289 - Review Results List for React - Optimization

nschubach opened the pull request:
## Description
Rewrite block to use functional components.  Rewrote tests using `@testing-library/react`

## Jira Ticket
- [TMEDIA-289](https://arcpublishing.atlassian.net/browse/TMEDIA-289)

## Acceptance Criteria
Optimize React Render methods

## Test Steps
- Add test steps a reviewer must complete to test this PR

0. Run `npm i` for additional packages around testing
1. Checkout this branch `git checkout TMEDIA-289-result-list-refactor`
2. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/results-list-block`
3. Create a new page and add the results-list-block to it
4. Configure the results-list-block as follows:
![image](https://user-images.githubusercontent.com/2287238/124684307-cec1bd80-de9c-11eb-8fa3-0ca4652ab133.png)
5. You should see it render the `/sara-carothers` articles 5 at a time.
6. Configure the results-list-block as follows:
![image](https://user-images.githubusercontent.com/2287238/124684661-a6868e80-de9d-11eb-947f-24ca93ceb5da.png)
7. You should see the homepage articles starting from the third article. (verify this by setting from to 0 and see the third)
8. Configure the following values and validate no button (I only have two stories for this, if you have more, increase size):
![image](https://user-images.githubusercontent.com/2287238/124686095-7bea0500-dea0-11eb-9135-716cfe57c3fa.png)
9. Page should preload 5 additional items, but not render them.  You can check this by viewing the network tab in debug and counting the number of items returned from the initial request.  Any press of the `See More` button will only pull the number of items equal to your size.
![image](https://user-images.githubusercontent.com/2287238/124686378-07639600-dea1-11eb-9020-d9fbbbf2fab2.png)
![image](https://user-images.githubusercontent.com/2287238/124686413-15b1b200-dea1-11eb-95a7-fa6691ff0716.png)
![image](https://user-images.githubusercontent.com/2287238/124686433-2104dd80-dea1-11eb-8ed8-1030b073a71b.png)

## Effect Of Changes
### Before
Pages would load content as requested, causing a slight delay to the user
Render passes (first and after hitting see more per @JackHowa's recommendation):
![image](https://user-images.githubusercontent.com/2287238/125114793-6d763600-e0b8-11eb-9baa-3bca6428fce9.png)
![image](https://user-images.githubusercontent.com/2287238/125114877-8979d780-e0b8-11eb-82e5-c8d9bc1d0530.png)
Lighthouse Result:
![image](https://user-images.githubusercontent.com/2287238/125114714-520b2b00-e0b8-11eb-838b-94eaff4e70fb.png)


### After
Pages should show additional content immediately and load more in the background
Render passes:
![image](https://user-images.githubusercontent.com/2287238/125113323-85e55100-e0b6-11eb-9f72-bc62712a3b9f.png)
![image](https://user-images.githubusercontent.com/2287238/125113400-9e556b80-e0b6-11eb-88f8-94f74331afa5.png)
Lighthouse Result:
![image](https://user-images.githubusercontent.com/2287238/125113236-60f0de00-e0b6-11eb-85d8-709df8ce9ef3.png)


## Dependencies or Side Effects
N/A

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [x] Confirmed this PR has unit test files
  - [x] Ran `npm run test`, made sure all tests are passing
  - [x] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [x] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist
_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

Commit:
TMEDIA-289 - Review Results List for React - Optimization

Commit:
Fix some bad math

Commit:
Move testing dependency to root package

nschubach committed:
Merge branch 'canary' into TMEDIA-289-result-list-refactor

matthewroach reviewed (changes requested):
With such a big change set I've been testing it as a whole and these are the following issues/changes that differ based on the changes made

* Lazy Load option is still requesting the data on the server side
* Results list is never rendered on the server side, it is now completely client side rendered
* The initial render is set the page's focus to the first element of the results list and breaking the focus order - Focus should only be manually set after the user clicks see more
* Focus isn't set on the first new item to be rendered after clicking see more button when you have an offset value set
* The see more button is delayed in showing after the next set of results are shown

matthewroach commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,326 @@
+        const size = requestedOffset === configuredOffset
+          ? configuredSize * 2
+          : configuredSize;
+        const from = requestedOffset === configuredOffset
```
Each `case` block works out the same size and offset. The only difference is the object retuned keys. We could reduce the duplication and just return in correct formats in the `case` blocks

nschubach replied:
fixed in the latest

Commit:
Fix SSR, focus, and lazy loading.

nschubach commented:
fixed - Lazy Load option is still requesting the data on the server side
fixed - The initial render is set the page's focus to the first element of the results list and breaking the focus order - Focus should only be manually set after the user clicks see more
fixed - Focus isn't set on the first new item to be rendered after clicking see more button when you have an offset value set
fixed - Results list is never rendered on the server side, it is now completely client side rendered
investigating - The see more button is delayed in showing after the next set of results are shown

nschubach commented:
I can update the visibility logic of the button to make it remain on screen if a current request is active, but that shifts the "change" to make the button remain on screen until it finds out if there is more content and makes it vanish at the end of the list.

Is this preferable?

matthewroach commented:
The button being removed and then shown quickly was something I noticed as different between current and the updates in this pull request.

Commit:
Gratuitous use of useState and memo to reduce renders

nschubach commented:
This changes the 'See More' button visibility functionality and introduces measures to prevent re-renders.

Commit:
Fix ref passing to use forwarfRef like I should.

Commit:
Tidy up the reference code and reduce param passing

matthewroach commented on blocks/results-list-block/features/results-list/default.jsx:
```diff
@@ -1,407 +1,30 @@
+import Results from './results';
+
+const ResultsList = () => {
+  const { customFields, isAdmin } = useFusionContext();
```
`customFields` is a prop of features in Fusion. Getting them from a fusion hook seems like an anti pattern, and also something that is not consistent with other blocks

nschubach replied:
I thought it was more dirty to grab different values from various places, but I'll change it back.  Makes testing easier...

matthewroach commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,344 @@
+    customFields,
+    deployment,
+    isAdmin,
+  } = useFusionContext();
```
`useFusionContext` is used in the parent component - would it not be more performant to get the items needed in the parent and pass them in as props?

nschubach replied:
I figured this was cached... thereby making the "performance" nearly minuscule but I can change it.  Makes testing easier...

JackHowa replied:
was this suggestion based on previous code style addressed @nschubach ?  We were talking about this being the pattern elsewhere

nschubach replied:
I addressed this by getting the config values form the parameters... regardless, I think it's a minor performance difference (if any since I'm still querying the context)

matthewroach commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,344 @@
+    locale,
+    primaryLogoAlt,
+    resizerURL,
+  }] = useState(getProperties(arcSite) || {});
```
Lifting this to the parent and passing as prop would remove the need for `useState`

nschubach replied:
Makes sense with the other changes.

JackHowa replied:
nice nice, so this was addressed

matthewroach commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,344 @@
+    resizerURL,
+  }] = useState(getProperties(arcSite) || {});
+
+  const [imageProps] = useState({
```
Lifting this to the parent and passing as prop would remove the need for `useState`

nschubach replied:
Makes sense with the other changes.

matthewroach commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,344 @@
+          placeholderResizedImageOptions={placeholderResizedImageOptions}
+        />
+      ))}
+      {(viewableElements.length < resultList?.content_elements.length
```
See more button is showing even when the results list is shorter than the custom field size value.
Example 
content source - story-feed-query 
quey - matthew
size - 50

There are only 2 results for query `matthew` - but See More button is still shown

nschubach replied:
Resolved in 5286b9a

Commit:
Fix focus issue and add test for focus check

Commit:
Fix focus bug

matthewroach commented:
Do we have before and after performance metrics on the changes in the pull request? For such a drastic change in approach of the block it would be something worth while finding out if we are making things better or worse

Commit:
Fix review comments, see more logic

nschubach committed:
Merge branch 'canary' into TMEDIA-289-result-list-refactor

matthewroach commented on blocks/results-list-block/features/results-list/default.jsx:
```diff
@@ -1,407 +1,50 @@
+      <HeadingSection>
+        <Results
+          customFields={customFields}
+          fusionContext={fusionContext}
```
You are passing the entire fusionContent object to the child. This a very large object that the child component only needs a handful of items from, we should only pass what is needed.

nschubach replied:
resolving

matthewroach commented on blocks/results-list-block/features/results-list/default.jsx:
```diff
@@ -1,407 +1,50 @@
+        <Results
+          customFields={customFields}
+          fusionContext={fusionContext}
+          fusionProperties={fusionProperties}
```
Same as above, we should only pass down what is needed, this object isn't as big as the content by could become very large for some clients

nschubach replied:
resolving

matthewroach commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,347 @@
+    locale,
+  } = fusionProperties;
+
+  const [configuredOffset] = useState(
```
What's the reason to use state to store everything in?

nschubach replied:
I explained more in the comment to Jack below, but it's more to prevent re-renders. I don't need them on the simple values and am checking in code to revert some of that useState.

JackHowa replied:
I'm not able to replicate the reason for dismissing this. I was able to replicate this suggestion helping https://github.com/WPMedia/fusion-news-theme-blocks/pull/967/files

nschubach replied:
I assume resolved since you close the branch and the code was removed?

JackHowa commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,347 @@
+    locale,
+  } = fusionProperties;
+
+  const [configuredOffset] = useState(
+    parseInt(contentConfigValues.offset, 10)
+    || parseInt(contentConfigValues.feedOffset, 10)
+    || parseInt(contentConfigValues.from, 10)
+    || 0,
+  );
+  const [configuredSize] = useState(
+    parseInt(contentConfigValues.size, 10)
+    || parseInt(contentConfigValues.feedSize, 10)
+    || 10,
+  );
```
is there a reason why these are not set as variables without `useState`? They appear to help derive state elsewhere

```suggestion
  const configuredOffset = parseInt(contentConfigValues.offset, 10)
    || parseInt(contentConfigValues.feedOffset, 10)
    || parseInt(contentConfigValues.from, 10)
    || 0;

  const configuredSize = parseInt(contentConfigValues.size, 10)
    || parseInt(contentConfigValues.feedSize, 10)
    || 10;
```

nschubach replied:
I'm removing those states... it was more of a knee jerk reaction to new items causing a re-render and I (wrongly) felt under the gun to get a working state quickly.

matthewroach commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,347 @@
+  }
+`;
+
+const ResultsItem = React.memo(React.forwardRef(({
```
Pulling this out to it's own file would make this file a bit more readable, and possibly allow us to in the future utilize this component in other list blocks

nschubach replied:
I was hoping you didn't notice that.  I definitely thought about this... I suppose since this story is carried on, I will undertake that and update/make the appropriate tests for it.

JackHowa replied:
so this is resolved, right?

JackHowa replied:
no, it doesn't seem like it is

nschubach replied:
I can if you would like.  It's going to involve pulling some tests and this into their own files... maybe an hour?

nschubach replied:
done and committed

JackHowa commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,347 @@
+
+  const serviceQueryPage = useCallback((requestedOffset = 0) => {
+    /*
+      This sets up a window view of the data starting from the initial index
+      to *twice* the size.  When clicking on the more button, we will render
+      the next size worth of items and load the next size from the server.
```
this is different behavior than what we had initially, right? Assuming so, this decision to load more data than is necessary will have the benefit of faster functionality later. But there's the tradeoff that this will slow down the first render. Was this part of the intention of the ticket?

nschubach replied:
Heh, this was the initial concern for me on the ticket.  Previously, the feeds that did not have a `next` property would "preload" the next set of results to determine if the see more button should be shown, then they were thrown away.  So I thought it would be a better use experience and less wasteful to preload the next set of results so when the user hit the button there was no delay in showing them the results and we could load the next set in the background.

nschubach replied:
done

JackHowa commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,347 @@
+    query: { raw_image_url: targetFallbackImage, respect_aspect_ratio: true },
+  });
+
+  const [isServerSideLazy] = useState(lazyLoad && isServerSide() && !isAdmin);
```
this also seems like a candidate for refactoring. not sure why this is necessary to save in state. I'd recommend this blog post on when and when not to sync state https://kentcdodds.com/blog/dont-sync-state-derive-it

JackHowa replied:
```suggestion
  const isServerSideLazy= lazyLoad && isServerSide() && !isAdmin;
```

JackHowa replied:
if you look elsewhere in the code, this similar functionality is not set as react state

nschubach replied:
Yep, see above.

JackHowa replied:
"Yep", you're saying you see this elsewhere in the code? 
![Screen Shot 2021-07-08 at 16 21 45](https://user-images.githubusercontent.com/5950956/124992399-81faf580-e008-11eb-995d-9f9c2d485b0a.png)

nschubach replied:
sorry, "yep" I am removing the useState.  Sorry for the confusion.

JackHowa commented on blocks/results-list-block/features/results-list/results/index.jsx:
```diff
@@ -0,0 +1,347 @@
+    }
+  }, [focalElement]);
+
+  const [promoElements] = useState(resolveDefaultPromoElements(customFields));
+  const [phrases] = useState(getTranslatedPhrases(locale || 'en'));
```
similarly here

nschubach replied:
CC: @matthewroach : This is the main "win" for useState.  These functions (at least one of them) returns an object.  That object is a new object each time the render is called, so it causes the memoized components (and useEffects) to re-render every time even thought he values within don't change.  Since the phrases and the promoElements are not likely to change between renders, making them "state" is a way to prevent re-rendering.

JackHowa replied:
so I'm not seeing any effect of phrases on clicking 'read more'. I have been seeing promoElements benefits in usestate for preventing re-renders. it looks like you're using the ref there. Couldn't we use a stable key for this to prevent react from re-rendering the preexisting list items? I can send you the json for the react profiler 

<img width="1138" alt="Screen Shot 2021-07-09 at 10 28 38" src="https://user-images.githubusercontent.com/5950956/125101907-588da880-e0a0-11eb-8746-5a79da8d3126.png">

JackHowa replied:
looks like each item is being passed all promoElements?

JackHowa replied:
I was able to not useState for promoElements if I didn't pass each item all elements. I separated out the individual props of showDate, etc

https://github.com/WPMedia/fusion-news-theme-blocks/pull/967/files

<img width="714" alt="Screen Shot 2021-07-09 at 10 38 36" src="https://user-images.githubusercontent.com/5950956/125103599-211ffb80-e0a2-11eb-83ea-c24163e2a324.png">

nschubach commented:
Added some Performance metrics... lighthouse I think it in margin of error... react tools shows difference of upfront loading on the new code.

JackHowa reviewed (changes requested):
First of, I applaud you for diving into this functionality. Using useMemo and ref can be a powerful tool for reducing render cycles and unnecessary re-rendering. I think the the `See More` functionality is less glitchy now. But I'm not sure if that was the intent of the ticket. When I looked at this ticket myself, I looked at reducing render cycles and re-rendering. The react profiler was helpful for that. This stuff can be hard but I'll try to look at it on a line-by-line basis.

Looks like there's a few places where there's unnecessary useState usages. I'd recommend looking at when to use useState vs setting a plain old variable. I included a good resource in one of the comments. I removed the flagged `useState`s and pushed an alternate [branch](https://github.com/WPMedia/fusion-news-theme-blocks/tree/TMEDIA-289-result-list-refactor-derive-state) with seemingly no ill effects in my testing. 

I see that useMemo was used. I know the ticket was about reducing unnecessary render cycles, which can be tough. Overall, I found one more render cycles in this feature branch (4) than in canary (3). It also seemed like some content was re-rendered in the optimization feature branch. Were you able to see this in the react profiler? Update: based on your screenshots, they don't seem to match mine? I'm curious what behavior you used to see that

Another goal of the ticket was to reduce dom nodes and improve ligthouse score. I was not able to find a consistent difference between this branch and canary. Was there something that was done to improve the performance score? 

It's hard for me to compare on a line-by-line basis as there have been file renaming. And I'm not sure if we have adequate testing resources (which have been changed in this pr) to start with, to do a refactor of this magnitude. I like to keep my tests constant if I make a sizable refactor to a component. 

canary

<img width="1284" alt="Screen Shot 2021-07-08 at 15 37 20" src="https://user-images.githubusercontent.com/5950956/124988607-d485e300-e003-11eb-8535-128ab4546d67.png">

this branch 

<img width="646" alt="Screen Shot 2021-07-08 at 15 22 18" src="https://user-images.githubusercontent.com/5950956/124988664-e2d3ff00-e003-11eb-8df8-c3aaac5fa45e.png">
<img width="1265" alt="Screen Shot 2021-07-08 at 15 22 05" src="https://user-images.githubusercontent.com/5950956/124988681-e7001c80-e003-11eb-8255-4e6d9c6bf1fb.png">

JackHowa commented:
what do you mean by this? @nschubach 

> react tools shows difference of upfront loading on the new code.

JackHowa commented on blocks/results-list-block/features/results-list/results/helpers.js:
```diff
@@ -0,0 +1,49 @@
+
+  if (!data) { return state; }
+
+  const oldItemIds = state.content_elements.map((element) => element._id);
+  const uniqueNewItems = data.content_elements
+    .filter((element) => !oldItemIds.includes(element._id));
```
is there fallback if there's no content_elements or an _id. I don't remember this functionality from before

nschubach replied:
You mean a fallback for elements without an _id?  From what I can tell, that's not possible?

nschubach replied:
As far as functionality.  It's just me being paranoid about duplicate records.

JackHowa replied:
ok this is new but I don't think it could necessarily be an issue

JackHowa commented on blocks/results-list-block/package.json:
```diff
@@ -36,5 +36,7 @@
   },
-  "gitHead": "83fb8d6685958d85ddbfcfbf01d9a5c864cd6c95"
+  "gitHead": "83fb8d6685958d85ddbfcfbf01d9a5c864cd6c95",
+  "devDependencies": {},
+  "dependencies": {}
```
dev and dependencies can be removed

nschubach replied:
np.  they were added when I added the testing to this module, but I had to move it up to the root package.json.

Commit:
Restrict context and properties passed to the component.  Removed unnecessary usstates.

nschubach commented:
> what do you mean by this? @nschubach
> 
> > react tools shows difference of upfront loading on the new code.

All I meant by that is, I was seeing the original code initially load an empty element and the changes and then rendered the list after the constructor fetch.  Interested in how you got only 3-4 renders though.

This was a 10 size, author feed render with new code:
![image](https://user-images.githubusercontent.com/2287238/124994161-6d702a80-e013-11eb-9db7-e7eb5545101f.png)

Compared to the same page in canary:
![image](https://user-images.githubusercontent.com/2287238/124994250-91cc0700-e013-11eb-97ee-d18bc405c760.png)

JackHowa commented:
@nschubach Sounds like the 24 renders you were seeing the screenshot is from the version of engine theme sdk from a couple weeks ago. let me know if you can replicate faster initial load and fewer render cycles. I'm not sure if this ticket would be best if it was refocused on improving the 'read more' functionality while maintaining the class component and the previous tests. Like I said, it's hard to evaluate a rewrite without the preexisting tests. Also, from my testing, the initial load is longer and with more more render cycles than canary. 

If this was me, I might start over and refocus the ticket on 'See more' functionality. Always good to measure performance after a refactor and evaluate if it was worthwhile. In my branch, with some of the suggestions, I didn't see huge improvements either. (cc @RobinGiannattasio )

RobinGiannattasio commented:
Hey all! Catching up a bit on this PR - it seems like maybe the best next step is to regroup on Monday with a meeting with everyone who is not on vacation (ENJOY IT MATTHEW) so we can more easily sync on the goal of the ticket (improved performance via metrics) vs the other findings here (the somewhat less than ideal behavior of the see more button originally). I'll send out a meeting invite.

I think until then, enjoy trivia in the next few minutes, and focus you work. on other tickets to help reset and clear the brains :D

Commit:
Implement flattened params per Jack's recommendation. Elevated some other params.

JackHowa reviewed (changes requested):
It looks like some of the default, fallback values are missing. I'm not sure if clients always have a content source defined for their blocks

in feature branch, I see a red warning and error boundary box when there's no content source: 

<img width="1421" alt="Screen Shot 2021-07-12 at 17 12 54" src="https://user-images.githubusercontent.com/5950956/125362436-87bc4800-e334-11eb-8cc2-adf5741b0d62.png">

in canary, no red warning box with no content source: 

<img width="1439" alt="Screen Shot 2021-07-12 at 17 13 21" src="https://user-images.githubusercontent.com/5950956/125362404-76733b80-e334-11eb-82de-62eb7ebc8641.png">

JackHowa commented:
Looks like there's also a memory leak of some sort that I'm seeing in the admin: 

with this content 
<img width="278" alt="Screen Shot 2021-07-12 at 17 21 16" src="https://user-images.githubusercontent.com/5950956/125363083-9bb47980-e335-11eb-9f4b-bbbfc9f25839.png">

in canary branch, no memory leak warning 

<img width="1431" alt="Screen Shot 2021-07-12 at 17 20 51" src="https://user-images.githubusercontent.com/5950956/125363103-a3741e00-e335-11eb-97aa-da045042f67e.png">


in feature branch, I see memory leak warning 

<img width="1091" alt="Screen Shot 2021-07-12 at 17 21 44" src="https://user-images.githubusercontent.com/5950956/125363134-b1c23a00-e335-11eb-99ed-ca3047b58ea0.png">

Commit:
Moved ResultItem to it's own file, moved testing to it's respective file, fixed up some other tests.

nschubach committed:
Merge branch 'canary' into TMEDIA-289-result-list-refactor

nschubach commented:
I don't see the red warning on mine...
![image](https://user-images.githubusercontent.com/2287238/125364259-27330800-e340-11eb-95c6-3414c037f5ad.png)

JackHowa commented:
<img width="1436" alt="Screen Shot 2021-07-12 at 17 39 02" src="https://user-images.githubusercontent.com/5950956/125364395-01096a00-e338-11eb-8c30-3c85b14aedad.png">

I updated the content config to be an ans story without setting the query or the offset

nschubach commented:
I can see the memory warning, but I have not identified anything that would cause it yet.

I did fix the red box issue by putting in a nullish coalesce on the properties so it properly falls back to the default values that it should.
```parseInt(contentConfigValues?.offset, 10)```

Will be updating that here in a minute.

nschubach commented:
Interesting, the useContent is triggering that warning...
![image](https://user-images.githubusercontent.com/2287238/125367238-c9a1ba00-e345-11eb-80ee-302ba264794c.png)

nschubach commented:
I might need to run that one up the line to fusion

Commit:
fix default value situation where no contentConfigValues exist

nschubach commented:
Per Jason Young I have opened a ticket for the warning: https://arcpublishing.atlassian.net/browse/PEN-2018

RobinGiannattasio reviewed (approved):
Hey hey! Making sure this keeps moving - I went through the comments and it seems everything has been addressed. As per our sync earlier this week, let's move this through to QA. If any other issues come up, we can create new tickets to track the work.

Thanks all!

nschubach committed:
Merge branch 'canary' into TMEDIA-289-result-list-refactor

JackHowa reviewed (commented):
approved to go to qa by robin

JackHowa reviewed (approved):
Robin approved for qa
